workflows-cli/src/linter/linter_labels.rs

```diff
@@ -0,0 +1,330 @@
+use serde::Deserialize;
+use serde_yaml::Value;
+use std::{
+    fmt::{self},
+    path::Path,
+    vec,
+};
+
+const RULES_YAML: &str = include_str!("../../../.workflows-labels.yaml");
+
+pub struct LabelChecker {
+    rules: Rules,
+}
+
+impl Linter for LabelChecker {
+    fn lint(target: &Path) -> Result<Vec<String>, String> {
+        let manifest = get_manifest(target)?;
+        let label_checker = LabelChecker::new()?;
+
+        let metadata = manifest
+            .as_mapping()
+            .ok_or("Manifest is not a mapping.")?
+            .get("metadata")
+            .ok_or("Workflow must have metadata")?;
+
+        let labels = metadata.get("labels").ok_or("Workflows must have labels")?;
+        let annotations = metadata
+            .get("annotations")
+            .ok_or("Workflows must have annotations")?;
+
+        label_checker.validate(labels, annotations)
+    }
+}
+
+impl LabelChecker {
+    fn build_rules() -> Result<Rules, String> {
+        serde_yaml::from_str(RULES_YAML).map_err(|e| format!("Failed to parse rules: {e}"))
+    }
+
+    fn new() -> Result<LabelChecker, String> {
+        let rules = Self::build_rules()?;
+        Ok(LabelChecker { rules })
+    }
+
+    fn validate(&self, labels: &Value, annotations: &Value) -> Result<Vec<String>, String> {
+        let mut errors = vec![];
+
+        for rule in &self.rules.critical {
+            if let Err(e) = rule.check(labels, annotations) {
+                errors.push(e);
+            }
+        }
+
+        Ok(errors)
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type")]
+enum CriticalRule {
+    #[serde(rename = "stem")]
+    Stem(Stem),
+    #[serde(rename = "match")]
+    Match(Match),
+}
+
+#[derive(Debug, Deserialize)]
+struct Rules {
+    critical: Vec<CriticalRule>,
+}
+
+#[derive(Debug, Deserialize)]
+struct Match {
+    key: String,
+    location: Location,
+}
+
+#[derive(Debug, Deserialize)]
+struct Stem {
+    stem: String,
+    options: Vec<String>,
+    location: Location,
+}
+
+#[derive(Debug, Deserialize)]
+enum Location {
+    #[serde(rename = "annotation")]
+    Annotation,
+    #[serde(rename = "label")]
+    Label,
+}
+
+impl fmt::Display for Location {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Annotation => write!(f, "annotations"),
+            Self::Label => write!(f, "labels"),
+        }
+    }
+}
+
+trait RuleChecker {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String>;
+}
+
+impl RuleChecker for CriticalRule {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String> {
+        match self {
+            CriticalRule::Match(m) => m.check(labels, annotations),
+            CriticalRule::Stem(s) => s.check(labels, annotations),
+        }
+    }
+}
+
+impl RuleChecker for Match {
+    fn check(&self, labels: &Value, annotations: &Value) -> Result<(), String> {
+        let exists = match self.location {
```

Comment:
This enforces that the key is present in the expected location.

Should a match rule enforce that the key is *not present* in the wrong location?